add region to multi-region cmek cluster (#14649)
cockroachdb/docs: cockroachcloud/managing-cmek.md

1. To start the process, contact {{ site.data.products.dedicated }} by reaching out to your account team, or [creating a support ticket](https://support.cockroachlabs.com/). You must provide your **Organization ID**, which you can find in your [Organization settings page](https://cockroachlabs.cloud/settings).

> They will enable the CMEK feature for your {{ site.data.products.db }} Organization.

Reviewer: ```suggestion
Your account team will enable the CMEK feature for your {{ site.data.products.db }} organization.
```

                        "key_spec": {
                            "type": "AWS_KMS",
                            "uri": "{id-of-key}",
                            "auth_principal": "{role-with-kms-access}"
                        }
                    },
                    {
                        "region": "us-central-1",
                        "key_spec": {
                            "type": "AWS_KMS",
                            "uri": "{id-of-another-key}",
                            "auth_principal": "{another-role-with-kms-access}"
                        }
                    }
                ],
                "hardware": {
                    "machine_type": "n2-standard-8"
                }
            }
        }
        ```
    - **GCP**:
        {% include_cached copy-clipboard.html %}
        ```json
        # cmek_gcp_config.json
        {
            "dedicated": {
                "region_nodes": {
                    "us-west1": 3,
                    "us-central1": 5
                },
                "cmek_region_specs": [
                    {
                        "region": "us-west1",
                        "key_spec": {

>                             "type": "GCP_KMS",

Reviewer 2: nit: the type should be GCP_CLOUD_KMS. yes, this is not parallel with AWS_KMS; IIRC we decided to use the official names of the products in each cloud (which is just KMS for AWS, and cloud KMS for GCP)

>                             "type": "GCP_KMS",

Author: ```suggestion
                            "type": "GCP_CLOUD_KMS",
```

                            "uri": "{id-of-key}",
                            "auth_principal": "{service-account-with-kms-access}"
                        }
                    },
                    {
                        "region": "us-central1",
                        "key_spec": {

>                             "type": "GCP_KMS",

Author: ```suggestion
                            "type": "GCP_CLOUD_KMS",
```

                            "uri": "{id-of-another-key}",
                            "auth_principal": "{another-service-account-with-kms-access}"
                        }
                    }
                ],
                "hardware": {
                    "machine_type": "n2-standard-8"
                }
            }
        }
        ```

> 1. Send the request as a `PATCH` to the cluster endpoint:

Reviewer: ```suggestion
1. Send the payload as a `PATCH` request to the cluster endpoint:
```

    {% include_cached copy-clipboard.html %}
    ~~~shell
    
    CLUSTER_ID= #{ your cluster ID }
    API_KEY= #{ your API key }
    curl --request PATCH \
      --url "https://cockroachlabs.cloud/api/v1/clusters/${CLUSTER_ID}" \
      --header "Authorization: Bearer ${API_KEY}" \
      --header 'content-type: application/json' \
      --data "@cmek_config.json"
    ~~~

>     ~~~txt

Author: ```suggestion
```

> 

Author: ```suggestion
```

>     ~~~

Author: ```suggestion
```

      --data "@cmek_config.json"
    ```

> ## Add a region to a multi-region CMEK-enabled cluster

Reviewer: I don't like this heading, because it's not yet a multi-region cluster until you add a second region (or in another way, every single-region cluster is just a multi-region cluster with only one region so far). Why not just "to a CMEK-enabled cluster"?

Author: fair point.

> To add a region to a multi-region cluster, update your cluster's region definitions using the CockroachDB Cloud API [Update Cluster](https://www.cockroachlabs.com/docs/api/cloud/v1#operation/CockroachCloud_UpdateCluster) endpoint.

Reviewer: ```suggestion
To add a region to a cluster that already has CMEK enabled, update your cluster's region definitions using the {{ site.data.products.db }} API [Update Cluster](https://www.cockroachlabs.com/docs/api/cloud/v1#operation/CockroachCloud_UpdateCluster) endpoint.
```

1. Construct the data payload JSON to update you region definitions.

>     The payload format depends on ccording to which cloud provider you use:

Author: ```suggestion
    The payload format depends on which cloud provider you use:
```

    - **AWS**:
        {% include_cached copy-clipboard.html %}
        ```json
        # cmek_config.json
        {
            "dedicated": {
                "region_nodes": {
                    "us-west1": 3,
                    "us-central1": 5
                },
                "cmek_region_specs": [
                    {

>                         "region": "us-west1",

Reviewer 2: nit: AWS regions have an extra dash before the number, like `us-west-1`

Author: thanks!

                        "key_spec": {
                            "type": "AWS_KMS",
                            "uri": "{id-of-key}",
                            "auth_principal": "{role-with-kms-access}"
                        }
                    },
                    {
                        "region": "us-central1",
                        "key_spec": {
                            "type": "AWS_KMS",
                            "uri": "{id-of-key-2}",
                            "auth_principal": "{another-role-with-kms-access}""
                        }
                    }
                ],
                "hardware": {
                    "machine_type": "n2-standard-8"
                }
            }
        }
        ```
    - **GCP**:

        {% include_cached copy-clipboard.html %}

>         ```json

Reviewer: Missing code here

Reviewer 2: This will look pretty much the same as the AWS example, but the region doesn't have the extra dash like I noted above (so `us-west1` is correct for GCP), and you'll have type `"type": "GCP_CLOUD_KMS"`

Author: thanks!

        ```

> 1. Send the request to the update cluster endpoint:

Reviewer: I can't spot an "update cluster" endpoint. I think the magic is that it's a `PATCH` request, but it's just the cluster's own endpoint.

Author: fair point.

    {% include_cached copy-clipboard.html %}
    ~~~shell
    
    CLUSTER_ID= #{ your cluster ID }
    API_KEY= #{ your API key }
    curl --request PATCH \
      --url "https://cockroachlabs.cloud/api/v1/clusters/${CLUSTER_ID}" \
      --header "Authorization: Bearer ${API_KEY}" \
      --header 'content-type: application/json' \
      --data "@cmek_config.json"
    ~~~

    ~~~txt

    ~~~

## Revoke CMEK for a cluster

Revoking access to the CMEK means disabling all encryption/decryption of data in your cluster, which means preventing reading and writing any data from or to your cluster. This likely implies a shutdown of your service, with significant business implications. This is a disaster mitigation tactic to be used only in a scenario involving a severe, business critical security breach.

> This can be done temporarily or permanently. This action is performed at the level of your cloud provider.

Reviewer: ```suggestion
Within your KMS platform, you can revoke access to the CMEK temporarily or permanently.
```
I thought we decided to go with "CMEK key" rather than "CMEK" for clarity, even though it's redundant like ATM Machine.

                        "region": "us-west1",
                        "key_spec": {
                            "type": "AWS_KMS",
                            "uri": "{id-of-key}",
                            "auth_principal": "{role-with-kms-access}"
                        }
                    },
                    {
                        "region": "us-central1",
                        "key_spec": {
                            "type": "AWS_KMS",
                            "uri": "{id-of-key-2}",

>                             "auth_principal": "{another-role-with-kms-access}""

Reviewer: ```suggestion
                            "auth_principal": "{another-role-with-kms-access}"
```
